taskchampion/src/server/cloud/aws.rs

```diff
@@ -0,0 +1,628 @@
+    fn put_and_get() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"foo").unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"foo".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn get_missing() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"data").unwrap();
+        svc.del(&pfx("testy")).unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del_missing() {
+        // Deleting an object that does not exist is not an error.
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        assert!(svc.del(&pfx("testy")).is_ok());
+    }
+
+    #[test]
+    fn list() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        let mut names: Vec<_> = (0..20).map(|i| pfx(&format!("pp-{i:02}"))).collect();
+        names.sort();
+        // Create 20 objects that will be listed.
+        for n in &names {
+            svc.put(n, b"data").unwrap();
+        }
+        // And another object that should not be included in the list.
+        svc.put(&pfx("xxx"), b"data").unwrap();
+
+        let got_objects: Vec<_> = svc.list(&pfx("pp-")).collect::<Result<_>>().unwrap();
+        let mut got_names: Vec<_> = got_objects.into_iter().map(|oi| oi.name).collect();
+        got_names.sort();
+        assert_eq!(
+            got_names
+                .iter()
+                .map(|b| String::from_utf8(b.to_vec()).unwrap())
+                .collect::<Vec<_>>(),
+            names
+                .iter()
+                .map(|b| String::from_utf8(b.to_vec()).unwrap())
+                .collect::<Vec<_>>()
+        );
+
+        // Clean up.
+        for n in got_names {
+            svc.del(&n).unwrap();
+        }
+        svc.del(&pfx("xxx")).unwrap();
+    }
+
+    #[test]
+    fn compare_and_swap_create() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        assert!(svc
+            .compare_and_swap(&pfx("testy"), None, b"bar".to_vec())
+            .unwrap());
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"bar".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn compare_and_swap_matches() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Create the existing file, with two generations.
+        svc.put(&pfx("testy"), b"foo1").unwrap();
```

Comment:
I don't understand your goal with the "two generations" here. How might the original write of that file impact the assertions?

Reply:
Oh, now that I look at your next comment, `compare_and_swap_mismatch` is checking that, so I'll leave the two tests with some better comments.